kong/tools/protojson.lua

```diff
@@ -0,0 +1,600 @@
+    struct.fields[ tostring( k ) ] = json_to_value_msg( item )
+  end
+
+  return struct
+end
+
+-- https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#listvalue
+-- https://github.com/protocolbuffers/protobuf/blob/master/python/google/protobuf/json_format.py#L693
+json_to_list_msg = function( v )
+  local list = { values = {} }
+
+  for _, item in ipairs(v or {}) do
+    table.insert( list.values, json_to_value_msg( item ) )
+  end
+
+  return list
+end
+
+-- https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#value
+-- https://github.com/protocolbuffers/protobuf/blob/master/python/google/protobuf/json_format.py#L675
+json_to_value_msg = function( v )
+  if type(v) == 'table' then
+    local c = 0
+
+    for _ in pairs(v) do c = c + 1 end
+
+    if #v == c then -- type = list
+      return { list_value = json_to_list_msg(v) }
+    else -- type = struct
+      return { struct_value = json_to_struct_msg(v) }
+    end
+  elseif type(v) == 'boolean' then
+    return { bool_value = v }
+  elseif type(v) == 'number' then
+    return { number_value = v }
+  elseif type(v) == 'string' then
+    return { string_value = v }
+  else
+    return { null_value = 0 }
+  end
+end
+
+--[[ API ]]------------------------------------------------------------------------
+
+--- Encodes output of `cjson.encode()` into bytestream
+-- @param msg table Representing json protobuf message
+-- @param msg_type Protobuf type of transformed message.
+-- @param path For error localisation [optional]
+function protojson:encode_from_json( msg_type, msg, path )
+  return pb.encode( msg_type, json_to_message( msg, msg_type, path ) )
+end
+
+--- Decode bytestream to to input for `cjson.encode()`
+-- @param msg Table representing json protobuf message
+-- @param msg_type Protobuf type of transformed message.
+-- @param path For error localisation [optional]
+function protojson:decode_to_json( msg_type, msg, path )
+  if ( emit_defaults ) then
+    pb.option( "use_default_values" )
+  else
+    pb.option( "no_default_values" )
+  end
+
+  if ( enum_as_name ) then
+    pb.option( "enum_as_name" )
+  else
+    pb.option( "enum_as_value" )
+  end
+
+  pb.option( "int64_as_number" )
+
+  return message_to_json( pb.decode( msg_type, msg ), msg_type, path )
+end
+
+--- Decode bytestream to to input for `cjson.encode()`
+-- @param options.use_proto_names_ Emit proto field name instead of lowerCamelCase name. Read both.
+-- @param options.enum_as_name_ Emit enum values as integers instead of strings. Read both.
+-- @param options.emit_defaults_ Emit fields with default values
+-- @param options.json_names_ Table containing json_name options of parsed fields
+function protojson:configure( options )
+  use_proto_names = options.use_proto_names or false
+  enum_as_name = options.enum_as_name or false
+  emit_defaults = options.emit_defaults or false
+  json_names = options.json_names or {}
+end
+
+return protojson
```

Comment:
We need a new line at EoF.

```suggestion
return protojson

```

Reply:
fixed: https://github.com/Kong/kong/pull/8297/commits/07c78f92ce6ed22707b67882f97adbeeb0265c39